kubernetes/updater/pkg/controller/updater.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"strings"
+
+	"github.com/docker/distribution/reference"
+	"github.com/gravitational/trace"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/gravitational/teleport/kubernetes/updater/pkg/img"
+	"github.com/gravitational/teleport/kubernetes/updater/pkg/maintenance"
+	"github.com/gravitational/teleport/kubernetes/updater/pkg/version"
+)
+
+type VersionUpdater struct {
+	versionGetter       version.Getter
+	imageValidators     img.Validators
+	maintenanceTriggers maintenance.Triggers
+	baseImage           reference.Named
+}
+
+// TODO: find a better name
+func (r *VersionUpdater) Do(ctx context.Context, obj v1.Object, currentVersion string) (img.NamedTaggedDigested, error) {
```

Comment:
Maybe `DoMaintenance`  will help, maybe not

```suggestion
func (r *VersionUpdater) DoMaintenance(ctx context.Context, obj v1.Object, currentVersion string) (img.NamedTaggedDigested, error) {
```

Reply:
What about
- `UpdateTeleportVersion`
- `UpdateTeleportVersionIfPermitted`
- `CheckMaintenanceGetVersionValidateImage`
- `CheckMaintenanceAndGetNewImage`

Reply:
It's not actually updating anything, is it? I'd just call it `GetTargetVersion` or something.